autorest/azure/devicetoken.go

```diff
@@ -0,0 +1,196 @@
+package azure
+
+/*
+  This file is largely based on rjw57/oauth2device's code, with the follow differences:
+   * scope -> resource, and only allow a single one
+   * receive "Message" in the DeviceCode struct and show it to users as the prompt
+   * azure-xplat-cli has the following behavior that this emulates:
+     - does not send client_secret during the token exchange
+     - sends resource again in the token exchange request
+*/
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest"
+)
+
+const (
+	// OAuthDeviceEndpoint is Azure's OAuth2 Device Flow Endpoint
+	OAuthDeviceEndpoint = "https://login.microsoftonline.com/common/oauth2/devicecode"
+	// OAuthTokenEndpoint is Azure's OAuth2 Token Endpoint
+	OAuthTokenEndpoint = "https://login.microsoftonline.com/common/oauth2/token"
+
+	// authAPIVersionQueryParamName is the name
+	authAPIVersionQueryParamName  = "api-version"
+	authAPIVersionQueryParamValue = "1.0"
+)
+
+var (
+	// ErrDeviceGeneric represents an unknown error from the token endpoint when using device flow
+	ErrDeviceGeneric = fmt.Errorf("Error while retrieving OAuth token: Unknown Error")
```

Comment:
we need to prefix all errors with `autorest/azure/devicetoken:` or `devicetoken:`